Highbar.Types/Either.cs

```diff
@@ -0,0 +1,365 @@
+    }
+
+    /// <summary>
+    /// Creates a <c>Right</c> from an arbitrary value.
+    /// </summary>
+    public static Either<SL, SR> Right<SL, SR>(SR right)
+    {
+      return new Right<SL, SR>(right);
+    }
+
+    /// <summary>
+    /// Alt <c>alt</c>.  Replace a <c>Left</c> with a <c>Right</c> of a value.
+    /// </summary>
+    public abstract Either<L, R> Alt(Either<L, R> other);
+
+    /// <summary>
+    /// Alt <c>alt</c>.  Replace a <c>Left</c> with a <c>Right</c> of a value produced by a <c>Func</c>.
+    /// </summary>
+    public abstract Either<L, R> Alt(Func<Either<L, R>> supplier);
+
+    /// <summary>
+    /// Alt <c>alt</c>.  Replace a <c>Left</c> with a <c>Right</c> of a value produced by a <c>Func</c>.
+    /// </summary>
+    public abstract Either<L, R> Alt(Func<L, Either<L, R>> other);
+
+    /// <summary>
+    /// Apply <c>ap</c>.  Applies the current <c>value</c> to the <c>value</c> of the <c>other</c>.
+    /// </summary>
+    public abstract Either<L, MR> Ap<MR>(Either<L, Func<R, MR>> other);
+
+    /// <summary>
+    /// Bifunctor <c>bimap</c>.  Maps either the <c>Left</c> or <c>Right</c> value without needing to know the state of the <c>Either</c>.
+    /// </summary>
+    public abstract Either<ML, MR> Bimap<ML, MR>(Func<L, ML> leftMapper, Func<R, MR> rightMapper);
+
+    /// <summary>
+    /// See Either#Chain(Func) chain.
+    /// </summary>
+    public Either<L, MR> Bind<MR>(Func<R, Either<L, MR>> mapper)
+    {
+      return Chain(mapper);
+    }
+
+    /// <summary>
+    /// Chain <c>chain</c> (a.k.a. <c>bind</c> or <c>flatMap</c>).  Takes a <c>Func</c> that accepts the <c>value</c> and returns an <c>Either</c> of the return value.
+    /// </summary>
+    public abstract Either<L, MR> Chain<MR>(Func<R, Either<L, MR>> mapper);
+
+    /// <summary>
+    /// See Either#alt(Either) alt.
+    /// </summary>
+    public Either<L, R> Coalesce(Either<L, R> other)
+    {
+      return Alt(other);
+    }
+
+    /// <summary>
+    /// Extend <c>extend</c>.  Takes a <c>Func</c> that accepts the <c>Either</c> and returns a <c>value</c>.
+    /// </summary>
+    public abstract Either<L, MR> Extend<MR>(Func<Either<L, R>, MR> mapper);
+
+    /// <summary>
+    /// Applies the <c>predicate</c> to the <c>Right</c> value.
+    /// </summary>
+    public abstract Maybe<R> Filter(Predicate<R> predicate);
+
+    /// <summary>
+    /// Applies the <c>predicate</c> to the <c>Right</c> value.
+    /// </summary>
+    public abstract Either<L, R> Filter(Predicate<R> predicate, L value);
+
+    /// <summary>
+    /// Applies the <c>predicate</c> to the <c>Right</c> value.
+    /// </summary>
+    public abstract Either<L, R> Filter(Predicate<R> predicate, Func<L> supplier);
+
+    /// <summary>
+    /// Bifunctor <c>first</c>.  Maps the <c>Left</c> value if the <c>instance</c> is a <c>Left</c>.
+    /// </summary>
+    public Either<ML, R> First<ML>(Func<L, ML> mapper)
+    {
+      return Bimap(mapper, i => i);
```

Comment:
Is there not a static `identity` function? If not, consider implementing on like in `Functions` class from the `functional-extras` library.